platform/darwin/src/MGLGeoJSONSource.mm

```diff
@@ -155,41 +171,47 @@ - (void)setGeoJSONData:(NSData *)geoJSONData
     }
     
     NSString *string = [[NSString alloc] initWithData:_geoJSONData encoding:NSUTF8StringEncoding];
-    const auto geojson = mapbox::geojson::parse(string.UTF8String).get<mapbox::geojson::feature_collection>();
+    const auto geojson = mapbox::geojson::parse(string.UTF8String);
     self.rawSource->setGeoJSON(geojson);
     
-    _features = MGLFeaturesFromMBGLFeatures(geojson);
+    self.shape = MGLShapeFromGeoJSON(geojson);
 }
 
-- (void)setURL:(NSURL *)URL
+- (void)setShape:(MGLShape *)shape
 {
-    _URL = URL;
-    
-    if (self.rawSource == NULL)
-    {
-        [self commonInit];
+    [self willChangeValueForKey:@"shape"];
+    _shape = shape;
+    [self didChangeValueForKey:@"shape"];
+    
+    NS_MUTABLE_ARRAY_OF(id <MGLFeature>) *features = [NSMutableArray array];
+    
+    if ([_shape isMemberOfClass:[MGLShapeCollection class]]) {
+        MGLShapeCollection *collection = (MGLShapeCollection *)_shape;
+        for (MGLShape *shape in collection.shapes) {
+            if ([shape conformsToProtocol:@protocol(MGLFeature)]) {
+                [features addObject:(id <MGLFeature>)shape];
```

Comment:
Would it be more efficient to compute the `features` property as soon as `shape` is set, or on demand when calling `-features`? Or perhaps a lazy-loading strategy is called for, similar to how I treated `overlayBounds` in #7251.